skimage/transform/hough_transform.py

```diff
@@ -343,9 +344,31 @@ def hough_circle_peaks(hspaces, radii, min_xdistance=1, min_ydistance=1,
     else:
         s = np.argsort(accum)
 
+    # For circles with centers too close, only keep the one with
+    # the highest peak
+    accum_sorted, cx_sorted, cy_sorted, r_sorted = \
+        accum[s][::-1], cx[s][::-1], cy[s][::-1], r[s][::-1]
+    accum = []
+    cx = []
+    cy = []
+    r = []
+    for ac, x, y, ra in zip(accum_sorted, cx_sorted, cy_sorted, r_sorted):
```

Comment:
could we execute this piece of code only if `min/max_xdistance>1`? I guess tt can take quite a long time for a large number of circles with the two loops (I haven't tried). Or could cythonize this part. I wonder how long it takes says for a 2000x2000 image with 1000 random circles?

Reply:
Another option could be to use a KDTree here for the x and y coordinates (one might have to add some jitter, I don't know from the top of my head if duplicate coordinates are allowed for KDTree), and use the `query_ball_point` method (https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.spatial.KDTree.query_ball_point.html#scipy.spatial.KDTree.query_ball_point) to keep a complexity of O(N) (with N the number of circles).

Reply:
Just to quickly chime in: in case this data structure is a good fit and you decide to use it I'd recommend [cKDTree](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.cKDTree.html) which is the faster version written in C. I've had great experiences (speed ups) with it so far! :)

Reply:
I think cKDTree supports duplicate coordinates and it will return duplicates as if queried.

Reply:
> could we execute this piece of code only if `min/max_xdistance>1`? I guess it can take quite a long time for a large number of circles with the two loops (I haven't tried). Or could cythonize this part. 

it is definitely helpful to skip picking when default values are used. and i agree that the two nested loops can led to O(N^2) complexity (N = number of peaks) and that this part of code should be optimized.

Reply:
> Another option could be to use a KDTree here for the x and y coordinates ...

I have tried using a cKDTree as suggested by @lagru and I believe the code now has better performance.
 
One subtile difference is that `query_ball_point` might still keep circles which centers are close in one dimension but far enough than specified distance in 2d plane. For example, two circles A, B with center (0, 0), (10, 10) should be considered close if min distance is 10 in both x and y directions. However, B will not be returned if we search A's neighbors within a distance of 10, i.e. sqrt(2) > 1.

Reply:
Oh you're right. We really made our life more complicated with having two parameters for the distances along `x` and `y`, but ok why not. One could still use the cKDTree with the largest distance times sqrt(2), and then test if the KDE neighbours should be eliminated or not.

Reply:
> with the largest distance times sqrt(2)

or np.hypot(dx, dy)? =)

Reply:
In my latest change, I scaled in both x and y dimension so we don't get an unnecessarily long list of neighbors.